### PR TITLE
Fix KafkaRebalance status when deprecated or unknown property is in spec

### PR DESCRIPTION
**Pull request by sknot-rh**
Signed-off-by: Stanislav Knot <sknot@redhat.com>

### Type of change
- Bugfix

### Description
KafkaRebalanceAbstractOperator in difference to other operators does not create a new status for each reconciliation. Instead of that, it reuses the previous one. In this PR, new status is built but data from the old one is used. 
Status like bellow does not appear anymore. Also removing non-valid conditions work now (e.g. unknown property is removed)

```
  - lastTransitionTime: 2021-02-02T11:34:16.034937Z
    message: 'Contains object at path spec with an unknown property: asd'
    reason: UnknownFields
    status: "True"
    type: Warning
  - lastTransitionTime: 2021-02-02T11:34:16.052363Z
    message: 'Contains object at path spec with an unknown property: asd'
    reason: UnknownFields
    status: "True"
    type: Warning
```

### Checklist
- [x] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Review by tomncooper — APPROVED** [review]

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java [hunk 1]
```diff
@@ -100,19 +100,30 @@ public static Condition buildRebalanceCondition(String type) {
...
         List<Condition> conditionList = new ArrayList<>();
         Condition condition = StatusUtils.buildCondition(type, conditionStatus, null);
         conditionList.add(condition);
 
         if (status.getConditions() != null) {
             status.getConditions().forEach(cond -> {
                 if ("UnknownFields".equals(cond.getReason()) || "DeprecatedFields".equals(cond.getReason())) {
-                    conditionList.add(cond);
+                    if (!conditionAlreadyPresent(conditionList, cond)) {
+                        conditionList.add(cond);
```
Comment (ppatierno) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:112]: just trying to understand, why we should not put an already existing condition. If the same condition is happening, we should have the same one but with a different timestamp to show its latest update.

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java [hunk 2]
```diff
@@ -100,19 +100,30 @@ public static Condition buildRebalanceCondition(String type) {
...
+        status.setConditions(new ArrayList<>(conditionList));
+    }
 
-        status.setConditions(conditionList);
+    private static boolean conditionAlreadyPresent(List<Condition> conditions, Condition condition) {
+        return conditions.stream().filter(cond -> {
+            return condition.getReason().equals(cond.getReason()) &&
+                    condition.getType().equals(cond.getType()) &&
+                    condition.getStatus().equals(cond.getStatus()) &&
+                    condition.getMessage().equals(cond.getMessage());
```
Comment (ppatierno) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:125]: isn't it enough to filter on type? That should be the "key", or?

Comment (sknot-rh) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:125]: Type is for example a `warning`. So using it as a key would result to having just one warning even in the case when there are more ones. I think more suitable one is `reason`.

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java (hunk 1 above)
Comment (sknot-rh) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:112]: It would result in this: https://gist.github.com/im-konge/b26e068fce4c5c1395717ef7c4a807ab

Comment (ppatierno) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:112]: how does it work today without warnings then?

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java (hunk 2 above)
Comment (ppatierno) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:125]: Yes maybe

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java (hunk 1 above)
Comment (sknot-rh) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:112]: Before #4291 there is nothing besides of Proposal status.

Comment (ppatierno) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:112]: this is part of a proposal status but we didn't have these warnings conditions and the condition was always just one, or? Now we are moving to have more conditions then? @tomncooper do I remember wrong?

Comment (sknot-rh) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:112]: Correct. The issue is described in https://github.com/strimzi/strimzi-kafka-operator/issues/3637

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java [hunk 3]
```diff
@@ -100,19 +100,24 @@ public static Condition buildRebalanceCondition(String type) {
...
         }
+        // condition contains a timestamp so using a set does not help
         List<Condition> conditionList = new ArrayList<>();
         Condition condition = StatusUtils.buildCondition(type, conditionStatus, null);
         conditionList.add(condition);
 
-        if (status.getConditions() != null) {
-            status.getConditions().forEach(cond -> {
-                if ("UnknownFields".equals(cond.getReason()) || "DeprecatedFields".equals(cond.getReason())) {
+        status.getConditions().forEach(cond -> {
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:108]: Why do you think we don't need the null check anymore?

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java [hunk 4]
```diff
@@ -100,19 +100,24 @@ public static Condition buildRebalanceCondition(String type) {
...
+            if ("UnknownFields".equals(cond.getReason()) || "DeprecatedFields".equals(cond.getReason())) {
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:109]: Should `DeprecatedObjects` be added here as well to this condition?

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java [hunk 5]
```diff
@@ -100,19 +100,24 @@ public static Condition buildRebalanceCondition(String type) {
...
+                if (!conditionAlreadyPresent(conditionList, cond)) {
+                    cond.setLastTransitionTime(iso8601Now());
                     conditionList.add(cond);
                 }
-            });
-        }
+            }
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:114]: Could you please explain what this code does? It is common to have the messages with the same reason present multiple times. For example:

```yaml
    - lastTransitionTime: "2021-02-15T13:05:24.039453Z"
      message: In API version kafka.strimzi.io/v1beta1 the metrics property at path
        spec.kafka.metrics has been deprecated, and should now be configured using
        spec.kafka.metricsConfig.
      reason: DeprecatedFields
      status: "True"
      type: Warning
    - lastTransitionTime: "2021-02-15T13:05:24.039453Z"
      message: In API version kafka.strimzi.io/v1beta1 the metrics property at path
        spec.zookeeper.metrics has been deprecated, and should now be configured using
        spec.zookeeper.metricsConfig.
      reason: DeprecatedFields
      status: "True"
      type: Warning
```

So I'm not entirely sure if what you are doing here is right.

**Comment by ppatierno** [comment]
I have no more comments than the ones from Jakub. I will wait for the answers there.

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java (hunk 5 above)
Comment (sknot-rh) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:114]: https://gist.github.com/im-konge/b26e068fce4c5c1395717ef7c4a807ab

Comment (scholzj) [reply operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:114]: That is without of context => so hard to understand the issue there. But I do not think it goes against my comment => I'm not arguing that you should have there the same warning 100 times. I'm saying that the Reason is not expected to be unique in general and should not be treated as such.

Yesterday when I checked this code in IDE it seemed to be used by KafkaRebalance only ... so maybe it never worked properly 🤷🏻‍♂️.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java [hunk 6]
```diff
@@ -382,18 +380,18 @@ private String rebalanceStateConditionType(KafkaRebalanceStatus status) {
...
+                return onNew(reconciliation, host, apiClient, rebalanceOptionsBuilder, kafkaRebalance);
             case PendingProposal:
                 return onPendingProposal(reconciliation, host, apiClient, kafkaRebalance, rebalanceAnnotation, rebalanceOptionsBuilder);
             case ProposalReady:
                 return onProposalReady(reconciliation, host, apiClient, kafkaRebalance, rebalanceAnnotation, rebalanceOptionsBuilder);
             case Rebalancing:
                 return onRebalancing(reconciliation, host, apiClient, kafkaRebalance, rebalanceAnnotation);
             case Stopped:
-                return onStop(reconciliation, host, apiClient, rebalanceAnnotation, rebalanceOptionsBuilder);
+                return onStop(reconciliation, host, apiClient, rebalanceAnnotation, rebalanceOptionsBuilder, kafkaRebalance);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:391]: Should we try to keep the same order of the parameters here? I see 4 methods hare which have already `reconciliation, host, apiClient, kafkaRebalance, ...`. So maybe you can insert it after the `apiClient` in the two new places as well?

**Review by scholzj — COMMENTED** [review]
One nit, but LGTM otherwise. Maybe we can also have some test for it?

File: api/src/test/resources/io/strimzi/api/kafka/model/047-Crd-kafkaconnector.yaml [hunk 7]
```diff
@@ -110,4 +110,9 @@ spec:
               tasksMax:
                 type: integer
                 description: The maximum number of tasks for the Kafka Connector.
+              topics:
+                type: array
+                items:
+                  type: string
+                description: The list of topics used by the Kafka Connector.
```
Comment (ppatierno) [inline api/src/test/resources/io/strimzi/api/kafka/model/047-Crd-kafkaconnector.yaml:117]: how is that related? a rebase problem?

Comment (sknot-rh) [reply api/src/test/resources/io/strimzi/api/kafka/model/047-Crd-kafkaconnector.yaml]: Yeah

**Review by ppatierno — APPROVED** [review]

**Review by scholzj — APPROVED** [review]
